patches/server/0971-Add-damage-source-wrapper.patch

```diff
@@ -0,0 +1,1534 @@
++                if (org.bukkit.craftbukkit.event.CraftEventFactory.handleNonLivingEntityDamageEvent(this, source, io.papermc.paper.entity.damageorigin.DamageOriginWrapper.toApi(source), amount)) { // Paper - Damage source wrapper
+                     return false;
+                 }
+                 // CraftBukkit end
++                this.killedBy = source; // Paper
+                 this.kill();
+                 return false;
+             } else if (!this.isInvulnerableTo(source) && (true || !this.invisible) && !this.isMarker()) { // CraftBukkit
+                 // CraftBukkit start
+-                if (org.bukkit.craftbukkit.event.CraftEventFactory.handleNonLivingEntityDamageEvent(this, source, amount, true, this.invisible)) {
++                if (org.bukkit.craftbukkit.event.CraftEventFactory.handleNonLivingEntityDamageEvent(this, source, io.papermc.paper.entity.damageorigin.DamageOriginWrapper.toApi(source), amount, true, this.invisible)) { // Paper - Damage source wrapper
+                     return false;
+                 }
+                 // CraftBukkit end
+                 if (source.is(DamageTypeTags.IS_EXPLOSION)) {
+                     this.brokenByAnything(source);
++                    this.killedBy = source; // Paper
+                     this.kill();
+                     return false;
+                 } else if (source.is(DamageTypeTags.IGNITES_ARMOR_STANDS)) {
+@@ -532,6 +535,7 @@ public class ArmorStand extends LivingEntity {
+                         if (source.isCreativePlayer()) {
+                             this.playBrokenSound();
+                             this.showBreakingParticles();
++                            this.killedBy = source; // Paper
+                             this.kill();
+                             return flag1;
+                         } else {
+@@ -598,6 +602,7 @@ public class ArmorStand extends LivingEntity {
+         f1 -= amount;
+         if (f1 <= 0.5F) {
+             this.brokenByAnything(damageSource);
++            this.killedBy = damageSource; // Paper
+             this.kill();
+         } else {
+             this.setHealth(f1);
+@@ -771,7 +776,7 @@ public class ArmorStand extends LivingEntity {
+ 
+     @Override
+     public void kill() {
+-        org.bukkit.event.entity.EntityDeathEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callEntityDeathEvent(this, drops); // CraftBukkit - call event // Paper - make cancellable
++        org.bukkit.event.entity.EntityDeathEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callEntityDeathEvent(this, drops, com.google.common.util.concurrent.Runnables.doNothing(), killedBy == null ? io.papermc.paper.entity.damageorigin.DamageOrigins.outOfWorld() : io.papermc.paper.entity.damageorigin.DamageOriginWrapper.toApi(killedBy)); // CraftBukkit - call event // Paper - make cancellable // Paper - Damage source wrapper
```

Comment:
Instead of storing a new field, I would maybe just make a new parameter for the kill method exclusive to the armorstand.

Where, if it's null, (which should be called by the default kill() implemention, it would return out of world.

Reply:
Good idea !